properties/extras/uid.py

```diff
@@ -0,0 +1,234 @@
+"""Classes for dealing with HasProperties instances with unique IDs"""
+import uuid
+
+from six import string_types, text_type
+
+from .. import base, basic, handlers, utils
+from ..base.instance import CLASS_TYPES
+
+
+class HasUID(base.HasProperties):
+    """HasUID is a HasProperties class that includes unique ID
+
+    Adding a UID to HasProperties allows serialization of more complex
+    structures, including recursive self-references. They are serialized
+    to a flat dictionary of UID/HasUID key/value pairs.
+    """
+
+    _REGISTRY = dict()
+    _INSTANCES = dict()
+
+    uid = basic.String(
+        'Unique identifier',
+        default=lambda: text_type(uuid.uuid4()),
+    )
+
+    def __init__(self, **kwargs):
+        super(HasUID, self).__init__(**kwargs)
+        self._INSTANCES[self.uid] = self
+
+    @handlers.validator('uid')
+    def _ensure_unique(self, change):
+        if self.uid == change['value']:
+            pass
+        elif change['value'] in self._INSTANCES:
+            raise utils.ValidationError(
+                message='Uid already used: {}'.format(change['value']),
+                reason='invalid',
+                prop=change['name'],
+                instance=self,
+            )
+        return True
+
+    @handlers.observer('uid')
+    def _update_instances(self, change):
+        self._INSTANCES.update({change['value']: self})
+
+    @classmethod
+    def validate_uid(cls, uid):
+        return True
+
+    @classmethod
+    def get_by_uid(cls, uid):
+        return cls._INSTANCES.get(uid)
+
+    def serialize(self, include_class=True, save_dynamic=False, **kwargs):
+        """Serialize nested HasUID instances to a flat dictionary
+
+        **Parameters**:
+
+        * **include_class** - If True (the default), the name of the class
+          will also be saved to the serialized dictionary under key
+          :code:`'__class__'`
+        * **save_dynamic** - If True, dynamic properties are written to
+          the serialized dict (default: False).
+        * You may also specify a **registry** - This is the flat dictionary
+          where UID/HasUID pairs are stored. By default, no registry need
+          be provided; a new dictionary will be created.
+        * Any other keyword arguments will be passed through to the Property
+          serializers.
+        """
+        registry = kwargs.pop('registry', None)
+        if registry is None:
+            registry = dict()
+        if not registry:
+            root = True
+            registry.update({'__uid__': self.uid})
```

Comment:
I guess I wonder if this is better as `__root__` or `__head__`, or something like that. It seems like they are _all_ UIDs?

Reply:
Yeah, makes sense. I'll change it.